### PR TITLE
Getrange v1

### PR DESCRIPTION
**Pull request by nemanja-boric-sociomantic**

### CODE REVIEW

**Comment by nemanja-boric-sociomantic** [comment]
Blocked on making it work on DLS node.

File: src/dlsproto/client/internal/SharedResources.d [hunk 1]
```diff
@@ -46,6 +49,22 @@ public final class SharedResources
 
     private FreeList!(ubyte[]) buffers;
 
+    /***************************************************************************
+
+        Pool of RequestEventDispatcher instances.
+
+    ***************************************************************************/
+
+    private FreeList!(RequestEventDispatcher) request_event_dispatchers;
```
Comment (gavin-norman-sociomantic) [inline src/dlsproto/client/internal/SharedResources.d:58]: For the future: It occurs to me that we might be able to adapt `RequestEventDispatcher` to accept `void[]`s to use internally, doing away with the need to have a separate pool of them. (I think the only thing it contains that should be reused is a couple of arrays.)

File: src/dlsproto/client/internal/SharedResources.d [hunk 2]
```diff
@@ -62,6 +81,29 @@ public final class SharedResources
 
     private Lzo lzo;
 
+    /***************************************************************************
+
+        A SharedResource instance is stored in the ConnectionSet as an
+        Object. This helper function safely casts from thsi Object to a
```
Comment (gavin-norman-sociomantic) [inline src/dlsproto/client/internal/SharedResources.d]: Typo thsi

File: src/dlsproto/client/mixins/NeoSupport.d [hunk 3]
```diff
@@ -19,7 +19,7 @@ import ocean.core.Tuple;
     Template to strip the notifier from the variadic arguments.
 
     Params:
-        TypeToErase = type to erase from options
```
Comment (gavin-norman-sociomantic) [inline src/dlsproto/client/mixins/NeoSupport.d:22]: Could you break out the whitespace changes into a separate commit?

File: src/dlsproto/client/request/internal/GetRange.d [hunk 4]
```diff
@@ -62,161 +56,43 @@ public struct GetRange
...
+    import swarm.neo.client.IRequestSet;
+    import dlsproto.client.internal.SharedResources;
 
     import ocean.io.select.protocol.generic.ErrnoIOException: IOError;
-    import dlsproto.client.internal.SuspendableRequest;
+
+    import ocean.transition;
+    mixin TypeofThis!();
+
+    mixin BatchController!(typeof(this), IController);
```
Comment (gavin-norman-sociomantic) [inline src/dlsproto/client/request/internal/GetRange.d]: Comment maybe? It's easy to miss this among the imports, otherwise.

File: src/dlsproto/client/request/internal/GetRange.d [hunk 5]
```diff
@@ -317,422 +160,631 @@ public struct GetRange
...
+        The fiber that waits for a batch of records to arrive and passes it to
+        the user, then sends the `Continue` message to the node, in a loop.
+        Handles suspending the request through the controller, for resuming
+        call `resume`. Calling `stop` makes this routine terminate after all
+        remaining records have been passed to the user.
 
     ***************************************************************************/
 
-    private State stateInitialising ( )
+    private class RecordStream
```
Comment (gavin-norman-sociomantic) [inline src/dlsproto/client/request/internal/GetRange.d]: I guess these could be scope classes.

File: src/dlsproto/client/request/internal/GetRange.d [hunk 6]
```diff
@@ -317,422 +160,631 @@ public struct GetRange
...
+        public void addRecords ( in void[] record_batch )
+        {
+            // append record_batch to *this.batch_buffer, which may or may not
+            // be empty.
+
+            if (!(*this.batch_buffer).length)
+                enableStomping(*this.batch_buffer);
+
+            // Append record_batch, then set this.remaining_batch to reference
+            // the remaining records. To void a dangling slice if
```
Comment (gavin-norman-sociomantic) [inline src/dlsproto/client/request/internal/GetRange.d:457]: Typo void.

File: src/dlsproto/client/request/internal/GetRange.d [hunk 7]
```diff
@@ -317,422 +160,631 @@ public struct GetRange
...
+                    if (this.outer.context.shared_working.suspendable_control.suspended)
+                    {
+                        yield_count = 0;
+                        this.suspendFiber(FiberSuspended.RequestSuspended);
+                    }
+
+                    this.passRecordToUser(
+                        *this.outer.conn.message_parser.getValue!(time_t)(
+                            this.remaining_batch),
+                        this.outer.conn.message_parser.getArray!(Const!(void))(
```
Comment (gavin-norman-sociomantic) [inline src/dlsproto/client/request/internal/GetRange.d:531]: Hmmm, do we want to do away with the batch compression? DMQ Consume never had it, but here we're losing something.

File: src/dlsproto/client/request/internal/GetRange.d (hunk 5 above)
Comment (gavin-norman-sociomantic) [reply src/dlsproto/client/request/internal/GetRange.d]: By the way, I presume that everything in here is completely generic (i.e. identical to DMQ Consume) except the extraction of records from the batch.

File: src/dlsproto/client/request/internal/GetRange.d [hunk 8]
```diff
@@ -317,422 +160,631 @@ public struct GetRange
...
+
+        **********************************************************************/
+
+        private void fiberMethod ()
+        {
+            while (true)
+            {
+                auto event = this.outer.request_event_dispatcher.nextEvent(
+                    this.fiber,
+                    Signal(ControllerSignal.Resume),
```
Comment (gavin-norman-sociomantic) [inline src/dlsproto/client/request/internal/GetRange.d:794]: What handles `controller.suspend`?

File: src/dlsproto/node/neo/request/GetRange.d [hunk 9]
```diff
@@ -266,291 +188,279 @@ public abstract scope class GetRangeProtocol_v0
...
-        Performs any logic needed to start reading records in the given
-        range.
-
-        Params:
-            low = lower range boundary
-            high = higher range boundary
-
-        Returns:
-            `true` if the range preparation was sucessfull
+        Request main loop. Pops records from the storage, sends them to the
```
Comment (gavin-norman-sociomantic) [inline src/dlsproto/node/neo/request/GetRange.d]: Pops? ;)

File: src/dlsproto/node/neo/request/GetRange.d [hunk 10]
```diff
@@ -266,291 +188,279 @@ public abstract scope class GetRangeProtocol_v0
...
 
-    /***************************************************************************
+            if (wait_for_data)
+            {
+                auto event = this.ed.nextEvent(ed.NextEventFlags.Receive);
 
-        Retrieve the next record from the channel, if available.
+                switch (event.active)
+                {
+                    // resumed from the node, resume the iteration
```
Comment (gavin-norman-sociomantic) [inline src/dlsproto/node/neo/request/GetRange.d]: "from the node" is a bit odd. This whole module is "the node", or?

File: src/dlsproto/node/neo/request/GetRange.d [hunk 11]
```diff
@@ -266,291 +188,279 @@ public abstract scope class GetRangeProtocol_v0
...
+            value = buffer to write the value into
+            wait_for_data = out parameter, will be set if the request should
+                            suspend until more data has arrived.
 
         Returns:
-            the state to change to according to `msg_type`.
-
-        Throws:
-            `ProtocolError` if `msg_type` is not a control message type.
+            `true` if there was a record in the channel, false if the channel is
```
Comment (gavin-norman-sociomantic) [inline src/dlsproto/node/neo/request/GetRange.d]: The doc of this method seems outdated. "channel is empty"?

**Review by gavin-norman-sociomantic — COMMENTED** [review]
Not feeling that mentally acute today. Just a few comments.

**Comment by nemanja-boric-sociomantic** [comment]
I'll address your comments, but just unblocking this as I've confirmed this works with the node (after 4f905e1 commit being applied).

File: src/dlsproto/client/internal/SharedResources.d (hunk 1 above)
Comment (nemanja-boric-sociomantic) [reply src/dlsproto/client/internal/SharedResources.d:58]: Yes. I will open the swarm issue about this.

File: src/dlsproto/client/request/internal/GetRange.d (hunk 7 above)
Comment (nemanja-boric-sociomantic) [reply src/dlsproto/client/request/internal/GetRange.d:531]: No, I think we should do the batch compression, thank you for reminding me.

File: src/dlsproto/client/request/internal/GetRange.d (hunk 5 above)
Comment (nemanja-boric-sociomantic) [reply src/dlsproto/client/request/internal/GetRange.d]:
```
./submodules/dlsproto/src/dlsproto/client/request/internal/GetRange.d(648): Error: variable dlsproto.client.request.internal.GetRange.GetRangeHandler.Reader.record_stream globals, statics, fields, ref and out parameters cannot be scope
```

> By the way, I presume that everything in here is completely generic (i.e. identical to DMQ Consume) except the extraction of records from the batch.

Yes, as the batches contains keys, and some few things (accepting the filter string and resuming when the new data arrives).

File: src/dlsproto/client/request/internal/GetRange.d (hunk 8 above)
Comment (nemanja-boric-sociomantic) [reply src/dlsproto/client/request/internal/GetRange.d:794]: Nothing, we're not expecting this signal asynchronously, but we're checking it before passing each record to the user:

https://github.com/sociomantic-tsunami/dlsproto/pull/39/files#diff-c6938f346e8e72b5984b4ebffc36d024R505

**Comment by nemanja-boric-sociomantic** [comment]
Updated with the comments addressed and with the batch compression.

**Comment by gavin-norman-sociomantic** [comment]
Ooh, I just remembered a trick David used recently in dmqproto that made reviewing _way_ easier when doing a complete request rewrite like this. `version(none)` the old handler (v0) and add the new handler (v1) at the top of the module. Then add a second commit that removes the versioned out code.

**Comment by nemanja-boric-sociomantic** [comment]
Ah, good idea, I will update.

**Comment by nemanja-boric-sociomantic** [comment]
Updated.

File: src/dlsproto/client/internal/SharedResources.d (hunk 1 above)
Comment (nemanja-boric-sociomantic) [reply src/dlsproto/client/internal/SharedResources.d:58]: https://github.com/sociomantic-tsunami/swarm/issues/186

File: src/dlsproto/node/neo/request/GetRange.d [hunk 12]
```diff
@@ -12,13 +12,527 @@
...
+            if (wait_for_data)
+            {
+                auto event = this.ed.nextEvent(ed.NextEventFlags.Receive);
+
+                switch (event.active)
+                {
+                    // Resumed by the storage engine, now ready to return value
+                    // with `getNextRecord`, so we try again.
+                    case event.active.resumed:
+                        continue main_loop;
```
Comment (david-eckardt-sociomantic) [inline src/dlsproto/node/neo/request/GetRange.d]: The `main_loop` label is not needed.

Comment (nemanja-boric-sociomantic) [reply src/dlsproto/node/neo/request/GetRange.d]: Yeah, it's _continue_ not _break_ :face_with_head_bandage:

Comment (nemanja-boric-sociomantic) [reply src/dlsproto/node/neo/request/GetRange.d]: Updated.

File: src/dlsproto/node/neo/request/GetRange.d [hunk 13]
```diff
@@ -12,13 +12,527 @@
...
+                            this.ed.NextEventFlags.Receive);
+
+                    if (event.active.received)
+                    {
+                        MessageType msg_type;
+                        this.parser.parseBody(event.received.payload, msg_type);
+
+                        if (msg_type == msg_type.Ack)
+                        {
+                            acked = true;
```
Comment (david-eckardt-sociomantic) [inline src/dlsproto/node/neo/request/GetRange.d]: Suggest replacing this variable with a `while (true)` loop and `break` here.

Comment (nemanja-boric-sociomantic) [reply src/dlsproto/node/neo/request/GetRange.d]: To be honest, I find the current approach more verbose (in a good way). I can still change it, if you insist.

**Review by gavin-norman-sociomantic — APPROVED** [review]
